fix: Avoid iOS block appender focus loop

dcalhoun opened the pull request:
## Related PRs

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5215

## What?

During programmatic Aztec focus events, only update the internal focused
element ref found within `TextInputState` and do not call the native focus
methods. 

## Why?

Fixes wordpress-mobile/WordPress-iOS#18783. Quickly adding multiple rich text
blocks on iOS resulted in a focus loop occurring between the rich text blocks.
The app would become unresponsive and eventually crash.

## How?

The focus callback triggered by Aztec-based programmatic focus events
can result in focus loops between rich text elements.

Android: This intentional no-op function prevents focus loops
originating when the native Aztec module programmatically focuses the
instance. The no-op is explicitly passed as an `onFocus` prop to avoid
future prop spreading from inadvertently introducing focus loops. The
user-facing focus of the element is handled by `onPress` instead.

See: https://github.com/wordpress-mobile/gutenberg-mobile/issues/302

iOS: Programmatic focus from the native Aztec module is required to
ensure the React-based `TextStateInput` ref is properly set when focus
is *returned* to an instance, e.g. dismissing a bottom sheet. If the ref
is not updated, attempts to dismiss the keyboard via the `ToolbarButton`
will fail.

See: https://github.com/wordpress-mobile/gutenberg-mobile/issues/702

The Android keyboard is, likely erroneously, already dismissed in the
contexts where programmatic focus may be required on iOS.

- https://github.com/WordPress/gutenberg/issues/28748
- https://github.com/WordPress/gutenberg/issues/29048
- https://github.com/wordpress-mobile/WordPress-Android/issues/16167

Programmatically swapping input focus creates an infinite loop if the
user taps a different input in between the programmatic focus and
the resulting update to the React Native TextInputState focused element
ref. To mitigate this, the below updates the focused element ref, but
does not call the native focus methods.

See: https://github.com/wordpress-mobile/WordPress-iOS/issues/18783

## Testing Instructions

<details><summary>Verify block appender loop fix</summary>

See wordpress-mobile/WordPress-iOS#18783.

1. Open a post in the editor.
1. Scroll to the bottom of the content.
1. Rapidly tap beneath the last block, which should add an empty Paragraph block.

</details>

<details><summary>Verify link picker keyboard dismissal</summary>

See https://github.com/wordpress-mobile/WordPress-iOS/issues/18783.

### iOS

1. Select a Paragraph block.
1. Present the Links UI (pressing the links button on the toolbar).
1. Dismiss the Links UI bottom-sheet.
1. Press the hide-keyboard button on the toolbar.
1. Check that the button works and the keyboard is dismissed.
1. Check that the block is now deselected.

### Android

> **Note** On Android, the keyboard is already hidden after dismissing the Links UI.

1. Select a Paragraph block.
1. Press the hide-keyboard button on the toolbar.
1. Check that the keyboard is dismissed.
1. Check that the block is now deselected.

</details>

<details><summary>Verify focus + selection conflict resolution</summary>

1. Open a post in the editor.
1. Add a Paragraph block. 
1. Add an Image block beneath the Paragraph block and set its media. 
1. Simultaneously tap both the Image and Paragraph block to select the Image block, but focus the Paragraph block. (It can take a few attempts to accomplish this race condition bug.) 
1. Verify tapping the keyboard hide button in the block toolbar dismisses the keyboard.[^1]

[^1]: This works on iOS, but fails on Android. 

</details>

<details><summary>Test general Writing Flow behavior</summary>

Given the foundational nature of this change, it feels appropriate to thoroughly test the writing experience for the block editor in general. Common [Writing Flow tests are documented](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases/gutenberg/writing-flow). 

</details>

## Screenshots or screencast

See:

- wordpress-mobile/WordPress-iOS#18783.
- https://github.com/wordpress-mobile/WordPress-iOS/issues/18783

dcalhoun committed:
fix: Avoid iOS block appender focus loop

The focus callback triggered by Aztec-based programmatic focus events
can result in focus loops between rich text elements.

Android: This intentional no-op function prevents focus loops
originating when the native Aztec module programmatically focuses the
instance. The no-op is explicitly passed as an `onFocus` prop to avoid
future prop spreading from inadvertently introducing focus loops. The
user-facing focus of the element is handled by `onPress` instead.

See: https://github.com/wordpress-mobile/gutenberg-mobile/issues/302

iOS: Programmatic focus from the native Aztec module is required to
ensure the React-based `TextStateInput` ref is properly set when focus
is *returned* to an instance, e.g. dismissing a bottom sheet. If the ref
is not updated, attempts to dismiss the keyboard via the `ToolbarButton`
will fail.

See: https://github.com/wordpress-mobile/gutenberg-mobile/issues/702

The Android keyboard is, likely erroneously, already dismissed in the
contexts where programmatic focus may be required on iOS.

- https://github.com/WordPress/gutenberg/issues/28748
- https://github.com/WordPress/gutenberg/issues/29048
- https://github.com/wordpress-mobile/WordPress-Android/issues/16167

Programmatic swapping focus from element to another often leads to focus
loops, only delegate the programmatic focus if there are no elements
focused.

See: https://github.com/wordpress-mobile/WordPress-iOS/issues/18783

dcalhoun commented on packages/react-native-aztec/src/AztecView.js:
```diff
@@ -231,13 +231,40 @@ class AztecView extends Component {
+			// - https://github.com/WordPress/gutenberg/issues/29048
+			// - https://github.com/wordpress-mobile/WordPress-Android/issues/16167
+			Platform.OS === 'ios' &&
+			// Programmatic swaping focus from element to another often leads to focus
+			// loops, only delegate the programmatic focus if there are no elements
+			// focused.
+			//
+			// See: https://github.com/wordpress-mobile/WordPress-iOS/issues/18783
+			! AztecInputState.getCurrentFocusedElement()
```
This is the key change. The rest is merely documentation for this complex context.

fluiddot replied:
`AztecInputState` already exposes a function for checking if any Aztec view is focused via [`isFocused`](https://github.com/WordPress/gutenberg/blob/3aa92a46adfaf052ffdeef1436cc648780d934b6/packages/react-native-aztec/src/AztecInputState.js#L50-L57), so probably we could use it instead of `getCurrentFocusedElement`:
```suggestion
			// Programmatic swaping focus from element to another often leads to focus
			// loops, only delegate the programmatic focus if there are no elements
			// focused.
			//
			// See: https://github.com/wordpress-mobile/WordPress-iOS/issues/18783
			! AztecInputState.isFocused()
```

dcalhoun committed:
docs: Add changelog entry

dcalhoun commented:
Hey @fluiddot. I requested your review as I noticed you worked on the Aztec component recently, specifically focus state management for the drag-and-drop feature. Please let me know if you'd prefer me request someone else, though. Thanks! 

Installable builds are available for [WPiOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/19467#issuecomment-1279391733) and [WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/pull/17342#issuecomment-1279357402).

fluiddot reviewed (changes requested):
Thanks @dcalhoun for addressing this issue 🙇 !

I've tested the changes and found an infinite focus loop when tapping several times (usually three or more times) beneath the last block (see attached video capture). I haven't explored further but looks like we might have here a race condition when two Aztec views notify the focus and blur events at the same time.

**NOTE:** I'm printing a `console.log` when the conditions that execute `_onPress` function are fulfilled.

https://user-images.githubusercontent.com/14905380/197745145-89c4fe10-8e75-49ce-a3fe-e752e9c020f0.mp4

dcalhoun committed:
fix: Programmatic Aztec input focus only updates internal ref

Programmatically swapping input focus creates an infinite loop if the
user taps a different input in between the programmatic focus and
the resulting update to the React Native TextInputState focused element
ref. To mitigate this, the Aztec now updates the focused element ref,
but does not call the native focus methods.

See: https://github.com/wordpress-mobile/WordPress-iOS/issues/18783

dcalhoun committed:
fix: Hide keyboard button forcibly dismisses the virtual keyboard

Due to race conditions between Aztec focus effects and React Native
gesture handlers, it is possible to select one block while placing focus
in a different block's text input. If the actively select block is not a
rich text field, tapping the dismiss keyboard button will not dismiss
the keyboard via clearing the selected block, as a non-rich-text block
does not trigger a blur event. So, this change explicitly dismisses the
keyboard rather than relying upon the selected block clear.

This is a hack. Ideally, the Aztec call loop is improved so that it is
not possible to select/focus more than one block.

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into fix/ios-block-appender-focus-loop

dcalhoun commented on packages/edit-post/src/components/header/header-toolbar/index.native.js:
```diff
@@ -185,6 +185,12 @@ export default compose( [
 			onHideKeyboard() {
 				clearSelectedBlock();
 				togglePostTitleSelection( false );
+				// Explicitly dismiss the keyboard for circumstances where Aztec race
+				// conditions lead to one block's rich text focused, but a different
+				// non-rich-text block selected. In this context, merely clearing block
+				// selection does not trigger a blur event and therefore will not
+				// dismiss the keyboard.
+				Keyboard.dismiss();
```
This change reintroduces an explicit keyboard dismissal that was removed in https://github.com/WordPress/gutenberg/commit/54323e94b843ad28e720ac1a095c8a981d287908 as it was deemed unnecessary. It _should_ be unnecessary, but there are race conditions between Aztec and React Native event handlers (e.g. block/rich text tap) that makes it possible to focus one block and select another when tapping the two near simultaneously. 

Explicitly dismissing the keyboard ensures that the user is able to close the keyboard when they find themselves in a state where a non-rich-text block is selected with the keyboard visible (the second image below).

| Rich Text | Non Rich Text |
| - | - |
| ![differing-focus-and-rich-text-block-select](https://user-images.githubusercontent.com/438664/199586443-64e96ee2-a85d-4f78-a4ef-7df78ca46ebe.jpeg) | ![differing-focus-and-non-rich-text-block-select](https://user-images.githubusercontent.com/438664/199586472-79d2d334-1cbe-4d02-91af-5dbe2b61368f.jpeg) |

This issue is not new in this PR, it is just more accessible now that focus loops are avoided on iOS. The change on this line fixes the focus + selection conflict on iOS. Unfortunately, it does not fix the issue on Android for an unknown reason currently.

dcalhoun reviewed (commented):
@fluiddot thank you for reviewing and testing this PR. I modified the approach taken to better mitigate focus loops like you experienced. 

It does seem like the latest approach works. Unfortunately, it does uncover an existing bug that I expand upon in a separate comment. The workaround for the uncovered bug only works for iOS. 

I will not be able to continue forward on this PR in the near future. Because of the state of this PR, I am uncertain if it is worth merging in its current state. It is arguably an improvement if it prevents focus loops on iOS. However, it is not the direct solution as it merely masks the underlying issue. 

If we'd like to leave this PR as-is for the time being without merging, that is understandable from my perspective. I am removing from myself as the assignee for the associated issues for the time being.

dcalhoun commented:
Incorporated into and superseded by https://github.com/WordPress/gutenberg/pull/53217.
